### PR TITLE
Implement Stream object

### LINKED ISSUE #66: Implement Stream object
Stream is an IO interface of VMO, which is useful in media device drivers.

Implement the following syscalls:
- [ ] zx_stream_create() - create a stream from a VMO
- [ ] zx_stream_readv() - read data from the stream at the current seek offset
- [ ] zx_stream_readv_at() - read data from the stream at a given offset
- [ ] zx_stream_writev() - write data to the stream at the current seek offset
- [ ] zx_stream_writev_at() - write data to the stream at a given offset
- [ ] zx_stream_seek() - modify the current seek offset of the stream

Pass all StreamTest in core-tests.

Reference: https://fuchsia.dev/fuchsia-src/reference/kernel_objects/stream

### PR DESCRIPTION
**Pull request by DeathWish5**
Finish stream object and pass all tests in core-stream.
If running this code in release mode, you will encounter a panic of page fault, because of a weird bug which makes my null pointer check fail, and in debug mode, everything works correctly. Although I tried for a long time, I can't fix it. If you are interested at this bug in release mode, contact me for more information.
The unsafe code in vm/stream.rs can be avoided by add some interface in VmObject and HAL, which damage the briefness, or do one more memory copy, which harm the efficiency. 

Resolve #66

### CODE REVIEW

File: zircon-object/src/vm/stream.rs [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+                    let target = content_size as isize + offset;
+                    if content_size as isize > 0 && target < 0 {
+                        return Err(ZxError::INVALID_ARGS);
+                    }
+                    *seek = target as usize;
+                }
+            }
+            _ => unreachable!(),
+        }
+        Ok(*seek)
```
Comment (wangrunji0408) [inline zircon-object/src/vm/stream.rs:243]: Is this ok?
```suggestion
        let mut seek = self.seek.lock();
        let origin = match seek_origin {
            SeekOrigin::Start => 0,
            SeekOrigin::Current => *seek,
            SeekOrigin::End => self.vmo.content_size(),
        };
        let (target, overflow) = (origin as isize).overflowing_add(offset);
        if overflow || target < 0 {
            return Err(ZxError::INVALID_ARGS);
        }
        *seek = target as usize;
        Ok(*seek)
```

File: zircon-object/src/vm/stream.rs [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+}
+
+#[repr(C)]
+#[derive(Default)]
+pub struct StreamInfo {
+    options: u32,
+    padding1: u32,
+    seek: u64,
+    content_size: u64,
+}
```
Comment (wangrunji0408) [inline zircon-object/src/vm/stream.rs:265]: Append docs here from Zircon source code.
(Because I'm confused with the concept of `content_size`...

**Review by wangrunji0408 — CHANGES_REQUESTED** [review]

File: zircon-syscall/src/stream.rs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        let handle = proc.add_handle(Handle::new(stream, rights));
+        out.write(handle)?;
+        Ok(())
+    }
+
+    pub fn sys_stream_writev(
+        &self,
+        handle_value: HandleValue,
+        options: u32,
+        user_bytes: UserInPtr<&[u8]>,
```
[the remark this answers is not in the document]
Comment (DeathWish5) [reply zircon-syscall/src/stream.rs]: `
#[repr(C)]
pub struct IoVec<T, P: Policy> {
    ptr: UserPtr<T, P>,
    len: usize,
}
`
You're right. With IoVec to limit memory layout, the bug was fixed.

File: zircon-syscall/src/stream.rs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+            if io_vec.as_ptr().is_null() {
+                return Err(ZxError::NOT_FOUND);
+            }
+            actual_count += stream.write(options, io_vec)?;
+        }
+        actual_count_ptr.write_if_not_null(actual_count)?;
+        Ok(())
+    }
+
+    pub fn sys_stream_writev_at(
```
[the remark this answers is not in the document]
Comment (DeathWish5) [reply zircon-syscall/src/stream.rs:58]: remove some duplicate code is easy, but I think merge them into one function is impossible.

File: zircon-syscall/src/stream.rs [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+            if io_vec.as_mut_ptr().is_null() {
+                return Err(ZxError::NOT_FOUND);
+            }
+            actual_count += stream.read(io_vec)?;
+        }
+        actual_count_ptr.write_if_not_null(actual_count)?;
+        Ok(())
+    }
+
+    pub fn sys_stream_readv_at(
```
[the remark this answers is not in the document]
Comment (DeathWish5) [reply zircon-syscall/src/stream.rs:127]: Remove some duplicate code is easy, but I think merge them into one function is impossible.

File: zircon-object/src/vm/stream.rs [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+impl From<usize> for SeekOrigin {
+    fn from(n: usize) -> Self {
+        match n {
+            0 => SeekOrigin::SeekOriginStart,
+            1 => SeekOrigin::SeekOriginCurrent,
+            2 => SeekOrigin::SeekOriginEnd,
+            _ => SeekOrigin::InvalidValue,
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (DeathWish5) [reply zircon-object/src/vm/stream.rs]: Done.

File: zircon-object/src/vm/stream.rs (hunk 1 above)
Comment (DeathWish5) [reply zircon-object/src/vm/stream.rs]: This one is ok.

```rust
let mut seek = self.seek.lock();
let origin: usize = match seek_origin {
    SeekOrigin::Start => 0,
    SeekOrigin::Current => *seek,
    SeekOrigin::End => self.vmo.content_size(),
};
if offset >= 0 {
    let (target, overflow) = origin.overflowing_add(offset as usize);
    if overflow {
        return Err(ZxError::INVALID_ARGS);
    }
    *seek = target;
} else {
    let target = origin as isize + offset;
    if origin as isize >= 0 && target < 0 {
        return Err(ZxError::INVALID_ARGS);
    }
    *seek = target as usize;
}
Ok(*seek)
```

The code suggested above will fail when :
* origin > MAX_ISIZE and offset < 0, which is correct but end up with target < 0
* origin > MAX_ISIZE and origin + offset > MAX_USIZE, which overflows but end up with correct

File: zircon-syscall/src/stream.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (wangrunji0408) [reply zircon-syscall/src/stream.rs:58]: You are right

**Review by wangrunji0408 — APPROVED** [review]

File: zircon-object/src/vm/stream.rs (hunk 1 above)
Comment (wangrunji0408) [reply zircon-object/src/vm/stream.rs]: 😵

**Pull request merged**